User permission manager

bright-starry-sky opened the pull request:
1,  moved `ClientSession` to common layer . and rename to `Session`.
2, Added Permission Manager .
3, Added Permission Check.

This PR depend on #1842 ，I will create more test cases after #1842 is done.

bright-starry-sky commented:
Ready to review. come on teachers. thanks.

dangleptr commented on src/common/permission/PermissionManager.cpp:
```diff
@@ -0,0 +1,214 @@
+
+// static
+bool PermissionManager::canReadSchemaData(session::Session *session) {
+    if (session->isGod()) {
```
why not put the check after `if (session->space() == -1) `

bright-starry-sky replied:
> why not put the check after `if (session->space() == -1)`

Good point.

bright-starry-sky replied:
> why not put the check after `if (session->space() == -1)`

fixed.

dangleptr commented on src/common/permission/PermissionManager.cpp:
```diff
@@ -0,0 +1,214 @@
+
+// static
+bool PermissionManager::canWriteSchema(session::Session *session) {
+    if (session->isGod()) {
```
ditto

dangleptr commented on src/common/permission/PermissionManager.cpp:
```diff
@@ -0,0 +1,214 @@
+    if (session->isGod()) {
+        return true;
+    }
+    if (session->space() == -1) {
```
ditto

dangleptr commented on src/common/session/Session.h:
```diff
@@ -0,0 +1,130 @@
+    int64_t           id_{0};
+    GraphSpaceID      space_{-1};
+    std::string       spaceName_;
+    std::string       user_;
```
maybe account is a better name.

bright-starry-sky replied:
> maybe account is a better name.

Good idea, changed.

dangleptr commented on src/meta/client/MetaClient.cpp:
```diff
@@ -1622,6 +1622,20 @@ const std::vector<HostAddr>& MetaClient::getAddresses() {
 }
 
+StatusOr<std::vector<nebula::cpp2::RoleItem>>
+MetaClient::getRolesByUser(const std::string& user) {
```
We should hold a cache for User->Role map

bright-starry-sky replied:
> We should hold a cache for User->Role map

Very good advice. changed the role map to cache.

dangleptr commented on src/common/session/Session.h:
```diff
@@ -0,0 +1,130 @@
+    }
+
+    bool isGod() const {
+        for (auto iter = roles_.begin(); iter != roles_.end(); iter++) {
```
Currently, we only have one user as GOD, that is root.

dangleptr commented on src/graph/ShowExecutor.cpp:
```diff
@@ -243,6 +244,14 @@ void ShowExecutor::showSpaces() {
         resp_->set_column_names(std::move(header));
 
         for (auto &space : retShowSpaces) {
+            if (FLAGS_enable_authorize) {
```
Maybe we could put the FLAGS_enable_authorize inside PermissionManager

dangleptr commented on src/meta/client/MetaClient.cpp:
```diff
@@ -1622,6 +1649,20 @@ const std::vector<HostAddr>& MetaClient::getAddresses() {
+}
+
+bool MetaClient::authenticationCheck(std::string account, std::string password) {
+    auto ret = authCheck(std::move(account), std::move(password)).get();
```
I think we should use cache for account->password too.

dangleptr commented on src/meta/processors/usersMan/AuthenticationProcessor.cpp:
```diff
@@ -198,19 +198,20 @@ void ListUsersProcessor::process(const cpp2::ListUsersReq& req) {
-        return;
+    auto spaceId = req.get_space_id();
+    /**
+     * Because the god role's space id is kDefaultSpaceId.
```
I am confused about this changes. 
"show roles" must be invoked under some space, right?  So why use defaultSpaceId here.

bright-starry-sky replied:
> I am confused about this changes.
> "show roles" must be invoked under some space, right? So why use defaultSpaceId here.

Good point.
Actual the "show roles" must be invoked under some space. show the use the defaultSpaceId  is redundant.
let me correct .

dangleptr commented on src/common/session/Session.h:
```diff
@@ -0,0 +1,130 @@
+        roles_.emplace(space, role);
+    }
+
+    uint64_t idleSeconds() const;
```
Where is the logic about taking care of expired session?

bright-starry-sky replied:
> Where is the logic about taking care of expired session?

No session expired implement yet. 
Because nebula1.0 is a simple SessionManager. This version will not support any management of sessions. 
And the `idleSeconds` keep it here because from original `SessionClient`  class.

bright-starry-sky commented:
Addressed dangleptr's comments as below : 
1,Cached user password via meta client; 
2,Check god by user name root;
3,Moved FLAGS_enable_authorize to PermissionManager;
4,Skip space check for kDefaultSpaceId when list roles

dangleptr commented on src/graph/PermissionCheck.cpp:
```diff
@@ -0,0 +1,139 @@
+        case Sentence::Kind::kDescribeSpace : {
+            /**
+             * Use space and Describe space are special operations.
+             * Permission checking needs to be done in their executor.
```
why? Miss the spaceId?

bright-starry-sky replied:
> why? Miss the spaceId?

Yes, Usually we need to get the space id from the session when check permission. 
We can't get the space ID form session when `Use Space` or `Describe space `, so we need get the space id in their executor.

dangleptr commented on src/graph/PermissionCheck.cpp:
```diff
@@ -0,0 +1,139 @@
+        case Sentence::Kind::kLimit :
+        case Sentence::Kind::KGroupBy :
+        case Sentence::Kind::kReturn : {
+            return permission::PermissionManager::canReadSchemaData(session);
```
canReadData?

bright-starry-sky replied:
> Maybe canReadSchemaOrData is a better name.

Good idea. Changed method name from `canReadSchemaData` to `canReadSchemaOrData`

panda-sheep commented on src/common/permission/PermissionManager.cpp:
```diff
@@ -0,0 +1,223 @@
+    }
+    bool havePermission = false;
+    switch (session->roleWithSpace(spaceId)) {
+        case session::Role::GOD :
```
What is the difference between  `session->isGod()`   and session::Role::GOD ？

bright-starry-sky replied:
> What is the difference between `session->isGod()` and session::Role::GOD ？

`session->isGod` just check if the user is god. 
`session->roleWithSpace(spaceId) `check what role of in this space.
`case session::Role::GOD` still there because the compilation requirement of switch block .

critical27 commented on src/graph/SequentialExecutor.cpp:
```diff
@@ -19,10 +19,21 @@ SequentialExecutor::SequentialExecutor(SequentialSentences *sentences,
+             * Skip special operations check at here. they are :
+             * kUse, kDescribeSpace, kRevoke and kGrant.
+             */
+            if (!PermissionCheck::permissionCheck(session, sentence)) {
```
What if the query is "use space test; sth else", can we work? But this is a quite usual.

bright-starry-sky replied:
> What if the query is "use space test; sth else", can we work? But this is a quite usual.

Good question. you are right . it's fail when "use space ; sth else" . 
because our permission check is in the preparation phase,  have not set the space id in the session yet in the preparation phase. so the second statement should be fail.
I originally intended to check permissions in each executor, but this would result in too much code clutter.

critical27 replied:
ok.

critical27 commented:
Clean code!

panda-sheep commented on src/common/session/Session.h:
```diff
@@ -0,0 +1,128 @@
+    std::string       account_;
+    time::Duration    idleDuration_;
+    /*
+     * map<space name, role>
```
map<space Id, role> ?

panda-sheep commented on src/common/session/Session.h:
```diff
@@ -0,0 +1,128 @@
+            }
+        }
+        return Role::INVALID_ROLE;
+    }
```
move `toRole ` to  `Session.cpp`  better ?
